// miyabi/dist/utils/agent-output.js
/**
 * AIエージェント向け出力ユーティリティ
 * Claude Codeなどのコーディングエージェントが機械可読な形式で結果を取得するための機能
 */
/**
 * Exit codes for agent consumption
 */
export var ExitCode;
(function (ExitCode) {
    ExitCode[ExitCode["SUCCESS"] = 0] = "SUCCESS";
    ExitCode[ExitCode["GENERAL_ERROR"] = 1] = "GENERAL_ERROR";
    ExitCode[ExitCode["CONFIG_ERROR"] = 2] = "CONFIG_ERROR";
    ExitCode[ExitCode["VALIDATION_ERROR"] = 3] = "VALIDATION_ERROR";
    ExitCode[ExitCode["NETWORK_ERROR"] = 4] = "NETWORK_ERROR";
    ExitCode[ExitCode["AUTH_ERROR"] = 5] = "AUTH_ERROR";
})(ExitCode || (ExitCode = {}));
/**
 * Check if JSON output is requested
 */
export function isJsonMode() {
    return process.argv.includes('--json') || process.env.MIYABI_JSON === '1';
}
/**
 * Check if auto-yes mode is enabled
 */
export function isAutoYes() {
    return process.argv.includes('--yes') ||
        process.argv.includes('-y') ||
        process.env.MIYABI_AUTO_YES === '1' ||
        process.env.CI === 'true';
}
/**
 * Check if verbose mode is enabled
 */
export function isVerbose() {
    return process.argv.includes('--verbose') ||
        process.argv.includes('-v') ||
        process.env.MIYABI_VERBOSE === '1';
}
/**
 * Output success result
 */
export function outputSuccess(data, message) {
    const output = {
        success: true,
        data,
        message,
        timestamp: new Date().toISOString(),
    };
    if (isJsonMode()) {
        console.log(JSON.stringify(output, null, 2));
    }
    else if (message) {
        console.log(message);
    }
}
/**
 * Output error and exit
 */
export function outputError(code, message, recoverable = true, suggestion, details) {
    const output = {
        success: false,
        error: {
            code,
            message,
            recoverable,
            suggestion,
            details,
        },
        timestamp: new Date().toISOString(),
    };
    if (isJsonMode()) {
        console.error(JSON.stringify(output, null, 2));
    }
    else {
        console.error(`❌ ${message}`);
        if (suggestion) {
            console.error(`\n💡 ${suggestion}`);
        }
    }
    const exitCode = getExitCodeFromError(code);
    process.exit(exitCode);
}
/**
 * Map error code to exit code
 */
function getExitCodeFromError(errorCode) {
    if (errorCode.includes('GITHUB_TOKEN') || errorCode.includes('CONFIG')) {
        return ExitCode.CONFIG_ERROR;
    }
    if (errorCode.includes('VALIDATION') || errorCode.includes('EXISTS')) {
        return ExitCode.VALIDATION_ERROR;
    }
    if (errorCode.includes('NETWORK') || errorCode.includes('API')) {
        return ExitCode.NETWORK_ERROR;
    }
    if (errorCode.includes('AUTH')) {
        return ExitCode.AUTH_ERROR;
    }
    return ExitCode.GENERAL_ERROR;
}
/**
 * Log verbose message
 */
export function logVerbose(message, data) {
    if (isVerbose() && !isJsonMode()) {
        console.log(`[VERBOSE] ${message}`);
        if (data) {
            console.log(JSON.stringify(data, null, 2));
        }
    }
}
/**
 * Log debug message
 */
export function logDebug(message, data) {
    if (process.env.MIYABI_DEBUG === '1' && !isJsonMode()) {
        console.log(`[DEBUG] ${message}`);
        if (data) {
            console.log(JSON.stringify(data, null, 2));
        }
    }
}
/**
 * Confirm action with user (or auto-yes in agent mode)
 */
export async function confirmAction(message, defaultValue = false) {
    if (isAutoYes()) {
        logVerbose(`Auto-confirmed: ${message}`);
        return true;
    }
    if (isJsonMode()) {
        // JSONモードでは確認できないのでエラー
        outputError('INTERACTIVE_NOT_SUPPORTED', 'Interactive prompts are not supported in JSON mode', true, 'Use --yes flag to auto-confirm');
    }
    // 対話型確認（既存のinquirerなどを使用）
    const inquirer = await import('inquirer');
    const { confirm } = await inquirer.default.prompt([
        {
            type: 'confirm',
            name: 'confirm',
            message,
            default: defaultValue,
        },
    ]);
    return confirm;
}
/**
 * Get GitHub token from environment or config
 */
export function getGitHubToken() {
    const token = process.env.GITHUB_TOKEN || process.env.GH_TOKEN;
    if (!token) {
        outputError('MISSING_GITHUB_TOKEN', 'GITHUB_TOKEN not found in environment', true, 'Set GITHUB_TOKEN environment variable: export GITHUB_TOKEN=ghp_xxx');
    }
    return token;
}
