// miyabi/dist/utils/cross-platform.js
/**
 * Cross-platform utilities for Miyabi CLI
 * Provides Windows/macOS/Linux compatibility
 */
import { execSync } from 'child_process';
import { platform } from 'os';
/**
 * Check if running on Windows
 */
export const isWindows = () => platform() === 'win32';
/**
 * Check if running on macOS
 */
export const isMacOS = () => platform() === 'darwin';
/**
 * Check if running on Linux
 */
export const isLinux = () => platform() === 'linux';
/**
 * Get the current platform name
 */
export const getPlatform = () => {
    const p = platform();
    if (p === 'win32')
        return 'windows';
    if (p === 'darwin')
        return 'macos';
    if (p === 'linux')
        return 'linux';
    return 'unknown';
};
/**
 * Execute a shell command with cross-platform compatibility
 * Automatically adds shell: true on Windows for proper command execution
 */
export function execCommand(command, options = {}) {
    const { silent, ...execOptions } = options;
    const opts = {
        encoding: 'utf-8',
        // Windows needs shell: true for commands to work properly
        shell: isWindows() ? 'cmd.exe' : undefined,
        stdio: silent ? ['pipe', 'pipe', 'ignore'] : execOptions.stdio,
        ...execOptions,
    };
    try {
        const result = execSync(command, opts);
        return result.toString().trim();
    }
    catch (error) {
        if (error instanceof Error && 'stdout' in error) {
            // Return stdout even on error (some commands return non-zero but have useful output)
            const stdout = error.stdout;
            if (stdout) {
                return stdout.toString().trim();
            }
        }
        throw error;
    }
}
/**
 * Execute a command and return the result without throwing on error
 */
export function execCommandSafe(command, options = {}) {
    try {
        const output = execCommand(command, options);
        return { success: true, output };
    }
    catch (error) {
        return {
            success: false,
            output: '',
            error: error instanceof Error ? error : new Error(String(error)),
        };
    }
}
/**
 * Check if a command is available on the system
 * Uses 'where' on Windows, 'command -v' on Unix
 */
export function isCommandAvailable(command) {
    try {
        if (isWindows()) {
            execSync(`where ${command}`, { stdio: 'ignore', shell: 'cmd.exe' });
        }
        else {
            execSync(`command -v ${command}`, { stdio: 'ignore', shell: '/bin/sh' });
        }
        return true;
    }
    catch {
        return false;
    }
}
/**
 * Get the path to a command
 * Returns null if the command is not found
 */
export function getCommandPath(command) {
    try {
        if (isWindows()) {
            const result = execSync(`where ${command}`, {
                encoding: 'utf-8',
                stdio: ['pipe', 'pipe', 'ignore'],
                shell: 'cmd.exe',
            });
            // 'where' can return multiple paths; take the first one
            return result.split('\n')[0].trim() || null;
        }
        else {
            const result = execSync(`command -v ${command}`, {
                encoding: 'utf-8',
                stdio: ['pipe', 'pipe', 'ignore'],
                shell: '/bin/sh',
            });
            return result.trim() || null;
        }
    }
    catch {
        return null;
    }
}
/**
 * Get the default shell for the current platform
 */
export function getDefaultShell() {
    if (isWindows()) {
        return process.env.COMSPEC || 'cmd.exe';
    }
    return process.env.SHELL || '/bin/sh';
}
/**
 * Normalize a path for display (convert backslashes to forward slashes)
 */
export function normalizePathForDisplay(path) {
    return path.replace(/\\/g, '/');
}
/**
 * Get spawn options with proper shell configuration
 */
export function getSpawnOptions(options = {}) {
    return {
        shell: isWindows() ? 'cmd.exe' : undefined,
        ...options,
    };
}
