// miyabi/dist/utils/human-in-the-loop.js
/**
 * Human-in-the-Loop Mode
 *
 * Steve Jobs Principle: "You've got to start with the customer experience
 * and work backward to the technology."
 *
 * This module provides approval gates for critical operations,
 * giving users control over autonomous agent actions.
 */
import chalk from 'chalk';
import inquirer from 'inquirer';
const DEFAULT_CONFIG = {
    approvalLevel: 'critical',
    autoApproveRisks: ['low'],
    notifyOnAutoApprove: true,
    timeoutSeconds: 300, // 5 minutes
};
const RISK_COLORS = {
    low: chalk.green,
    medium: chalk.yellow,
    high: chalk.red,
    critical: chalk.bgRed.white,
};
const RISK_ICONS = {
    low: 'â—‹',
    medium: 'â—',
    high: 'â—',
    critical: 'âš ',
};
/**
 * Human-in-the-Loop Manager
 */
export class HumanInTheLoop {
    config;
    history = [];
    isInteractive;
    constructor(config = {}) {
        this.config = { ...DEFAULT_CONFIG, ...config };
        this.isInteractive = process.stdin.isTTY && process.stdout.isTTY;
    }
    /**
     * Check if approval is required for a given gate
     */
    requiresApproval(gate) {
        if (this.config.approvalLevel === 'auto') {
            return false;
        }
        if (this.config.approvalLevel === 'all') {
            return true;
        }
        // 'critical' level - only require approval for high/critical risks
        return gate.risk === 'high' || gate.risk === 'critical';
    }
    /**
     * Format gate details for display
     */
    formatGateDetails(gate) {
        const lines = [];
        const riskColor = RISK_COLORS[gate.risk];
        const riskIcon = RISK_ICONS[gate.risk];
        lines.push('');
        lines.push(chalk.yellow.bold('â”'.repeat(60)));
        lines.push(chalk.yellow.bold('  ðŸš¨ Human Approval Required'));
        lines.push(chalk.yellow.bold('â”'.repeat(60)));
        lines.push('');
        lines.push(`  ${chalk.white.bold('Action:')} ${gate.action}`);
        lines.push(`  ${chalk.white.bold('Description:')} ${gate.description}`);
        lines.push(`  ${chalk.white.bold('Risk Level:')} ${riskColor(`${riskIcon} ${gate.risk.toUpperCase()}`)}`);
        if (gate.details) {
            lines.push('');
            lines.push(chalk.white.bold('  Details:'));
            for (const [key, value] of Object.entries(gate.details)) {
                lines.push(`    ${chalk.gray(key)}: ${chalk.white(String(value))}`);
            }
        }
        lines.push('');
        return lines;
    }
    /**
     * Request approval from user
     */
    async requestApproval(gate) {
        // Check if approval is needed
        if (!this.requiresApproval(gate)) {
            // Auto-approve low risk actions
            if (this.config.notifyOnAutoApprove && !this.config.autoApproveRisks.includes(gate.risk)) {
                console.log(chalk.gray(`  âš¡ Auto-approved: ${gate.action}`));
            }
            const result = {
                approved: true,
                reason: 'Auto-approved based on configuration',
                timestamp: new Date(),
            };
            this.history.push({ gate, result });
            return result;
        }
        // Display gate information
        console.log(this.formatGateDetails(gate).join('\n'));
        // Non-interactive mode - use suggested action or reject
        if (!this.isInteractive) {
            const approved = gate.suggestedAction === 'approve';
            const result = {
                approved,
                reason: approved ? 'Auto-approved (non-interactive mode)' : 'Rejected (non-interactive mode)',
                timestamp: new Date(),
            };
            this.history.push({ gate, result });
            return result;
        }
        // Interactive approval
        const { decision } = await inquirer.prompt([
            {
                type: 'list',
                name: 'decision',
                message: 'What would you like to do?',
                choices: [
                    { name: chalk.green('âœ“ Approve') + ' - Proceed with this action', value: 'approve' },
                    { name: chalk.red('âœ— Reject') + ' - Skip this action', value: 'reject' },
                    { name: chalk.yellow('â„¹ More Info') + ' - Show detailed information', value: 'info' },
                    { name: chalk.gray('â¸ Pause') + ' - Pause execution', value: 'pause' },
                ],
                default: gate.suggestedAction === 'approve' ? 'approve' : 'reject',
            },
        ]);
        if (decision === 'info') {
            // Show more details and ask again
            console.log(chalk.white.bold('\n  Additional Information:'));
            console.log(chalk.gray(`  Gate ID: ${gate.id}`));
            console.log(chalk.gray(`  Created: ${new Date().toISOString()}`));
            if (gate.details) {
                console.log(chalk.gray(`  Full Details: ${JSON.stringify(gate.details, null, 2)}`));
            }
            return this.requestApproval(gate);
        }
        if (decision === 'pause') {
            console.log(chalk.yellow('\n  â¸ Execution paused. Press Enter to continue...'));
            await inquirer.prompt([{ type: 'input', name: 'continue', message: '' }]);
            return this.requestApproval(gate);
        }
        let reason;
        if (decision === 'reject') {
            const { rejectReason } = await inquirer.prompt([
                {
                    type: 'input',
                    name: 'rejectReason',
                    message: 'Reason for rejection (optional):',
                    default: '',
                },
            ]);
            reason = rejectReason || 'User rejected';
        }
        const result = {
            approved: decision === 'approve',
            reason: decision === 'approve' ? 'User approved' : reason,
            timestamp: new Date(),
        };
        this.history.push({ gate, result });
        if (result.approved) {
            console.log(chalk.green('\n  âœ“ Approved. Proceeding...\n'));
        }
        else {
            console.log(chalk.red('\n  âœ— Rejected. Skipping this action.\n'));
        }
        return result;
    }
    /**
     * Create approval gates for common operations
     */
    static createGate(type, details = {}) {
        const gates = {
            deploy: {
                action: 'Deploy to Production',
                description: 'Deploy changes to production environment',
                risk: 'high',
                suggestedAction: 'approve',
            },
            merge: {
                action: 'Merge Pull Request',
                description: 'Merge code changes into main branch',
                risk: 'medium',
                suggestedAction: 'approve',
            },
            delete: {
                action: 'Delete Resources',
                description: 'Delete files, branches, or other resources',
                risk: 'high',
                suggestedAction: 'reject',
            },
            publish: {
                action: 'Publish Package',
                description: 'Publish package to npm registry',
                risk: 'critical',
                suggestedAction: 'approve',
            },
            execute: {
                action: 'Execute Command',
                description: 'Run a shell command',
                risk: 'medium',
                suggestedAction: 'approve',
            },
        };
        const baseGate = gates[type] || {
            action: 'Unknown Action',
            description: 'An unknown action requires approval',
            risk: 'high',
        };
        return {
            id: `${type}-${Date.now()}`,
            ...baseGate,
            details,
        };
    }
    /**
     * Get approval history
     */
    getHistory() {
        return [...this.history];
    }
    /**
     * Get summary of approvals
     */
    getSummary() {
        const approved = this.history.filter(h => h.result.approved).length;
        return {
            approved,
            rejected: this.history.length - approved,
            total: this.history.length,
        };
    }
    /**
     * Update configuration
     */
    setConfig(config) {
        this.config = { ...this.config, ...config };
    }
    /**
     * Get current configuration
     */
    getConfig() {
        return { ...this.config };
    }
}
/**
 * Quick approval helper for common use cases
 */
export async function requireApproval(action, description, risk = 'medium') {
    const hitl = new HumanInTheLoop();
    const gate = {
        id: `quick-${Date.now()}`,
        action,
        description,
        risk,
        suggestedAction: 'approve',
    };
    const result = await hitl.requestApproval(gate);
    return result.approved;
}
/**
 * Decorator for functions that require approval
 */
export function withApproval(action, description, risk = 'medium') {
    return function (_target, _propertyKey, descriptor) {
        const originalMethod = descriptor.value;
        if (!originalMethod)
            return descriptor;
        descriptor.value = async function (...args) {
            const approved = await requireApproval(action, description, risk);
            if (!approved) {
                throw new Error(`Action rejected: ${action}`);
            }
            return originalMethod.apply(this, args);
        };
        return descriptor;
    };
}
